src/util/parented_ptr.h

```diff
@@ -0,0 +1,78 @@
+#ifndef PARENTED_POINTER_H
+#define PARENTED_POINTER_H
+
+#include <memory>
+#include "assert.h"
+
+
+/**
+ * Use this wrapper class to clearly represent a pointer that is owned by the QT object tree.
+ * Objects which both derive from QObject AND have a parent object, have their lifetime governed by the QT object tree, 
+ * and thus pointers to such objects do not need to be deleted when they go  out of scope.
+**/
+template <typename T>
+class parented_ptr {
+  public:
+    explicit parented_ptr(T* t) : m_pPtr(t) {
+        // use C style assert here as it usefully prints the type T
+        assert(t->parent() != nullptr);
+    }
+
+    parented_ptr() : m_pPtr(nullptr) {}
+
+    ~parented_ptr() = default;
+
+    T* get() const {
+        return m_pPtr;
```

Comment:
> If we use get() here as well, it looks similar, but the whole new pointer does not have the ownership

It kind of does have ownership (by proxy) to the object tree. We could even add a `release()` function that removes the parent and returns the raw pointer.

Also, it's highly preferable to keep the interface consistent with the other std pointer wrappers to simplify switching between them. I imagine it may be common to switch from shared to parented, and it will be unnecessarily annoying if you have to go and change all the .get() calls to something else.